Documentation/en-us/TestUsingTestDoubles.md

```diff
@@ -0,0 +1,131 @@
+
+Here's one example: BTests tests B, which uses A. Now bugs in A cause BTests to fail, even though B is fine! 
+All the failures are confusing--it's hard to answer the question "what's broken?"
+
+To avoid this problem, we can use a stand-in for A in Btests. In this case, we call stand-in for A `A' class`.
+
+![](https://github.com/Quick/Assets/blob/master/Screenshots/TestUsingMock_BusesAmock.png)
+
+`A' class` has same methods--however implementation might be different, and properties as `A Class`. So we can replace `A Class` with `A' Class` in BTests.
+
+Objects like `A' Class` are called 'test doubles'. 'test doubles' are used as stand-in for tests.
+There are some kinds of 'test doubles'.
+
+- Mock object: used for receiving output from test class
+- Stub object: used for providing input to test class
+- Fake object: behaves similar as original class
+
+Here, we explain using `Mock` next section.
+
+## Mock
+
+Mock is object(class, struct) used for verifying test object works with other objects from point of view of test target's output.
+
+### Writing test with Mock in Swift
+
+#### Sample app
+
+For example, we show an app which retrieves data from internet.
+
+* Display data from internet in ArticleViewController
+* Custom class implements DataProviderProtocol is responsible for getting data.
+
+DataProviderProtocol is defined as follows,
+
+```swift
+// Swift
+protocol DataProviderProtocol: class {
+    func fetch(callback: (data: String) -> Void)
+}
+```
+`fetch()` gets data from internet and output data with callback block.
+
+Here's DataProvider which implements DataProviderProtocol.
+
+```swift
+// Swift
+class DataProvider: NSObject, DataProviderProtocol {
+    func fetch(callback: (data: String) -> Void) {
+        let url  = NSURL(string: "http://example.com/")!
+        let session = NSURLSession(configuration: NSURLSessionConfiguration.defaultSessionConfiguration())
+        let task    = session.dataTaskWithURL(url, completionHandler: {
+            (data, resp, err) in
+            let string = NSString(data:data!, encoding:NSUTF8StringEncoding) as! String
+            callback(data: string)
+        })
+        task.resume()
+    }
+}
+```
+In our scenario, `fetch()` is called in `viewDidLoad()` of ViewController.
+
+Code is below.
+
+```swift
+// Swift
+class ViewController: UIViewController {
+    @IBOutlet weak var resultLabel: UILabel!
+    private var dataProvider: DataProviderProtocol?
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+        dataProvider = dataProvider ?? DataProvider()
+
+        dataProvider?.fetch({ [unowned self] (data) -> Void in
+            self.resultLabel.text = data
+        })
+    }
+}
+```
+
+#### Writing test with Mock of DataProviderProtocol
+
+ArticleViewController depends on DataProviderProtocol.
+Here, create Mock which inherits DataProviderProtocol in Test Targets in order to test ViewController.
```

Comment:
Not to keep mentioning the same point, but I think this line is a good example of what I mean. If I were a reader who didn't know what mocks were, I would ask "why should I create a `MockDataProvider`? Why are you asking me to write this code?"

The actual reasons to write the mock are compelling:

1. Your tests will run much faster.
2. They will run correctly whether you're connected to the Internet or not.
3. They will not fail because of errors in your actual `DataProvider` (i.e.: test isolation).